[command] refactor evaluate impl

fenrir-z opened the pull request.

Commit:
move

Commit:
update

Commit:
update

Commit:
update

Commit:
update

Commit:
add test case for det_eval_voc

Commit:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
get_ious_array -> get_iou_thrs_array

phoenix-xhuang commented on ymir/command/mir/commands/evaluate.py:
```diff
@@ -92,4 +96,10 @@ def bind_to_subparsers(subparsers: argparse._SubParsersAction, parent_parser: ar
+                                     required=False,
+                                     type=str,
+                                     default='',
+                                     help='class names in mAP cauculate, keep empty to use all classes in gt')
```
规则不对，要是不提供就是用pred里绑定的模型类别

phoenix-xhuang replied:
如果不想在这个版本加模型绑定信息，就加个fixme

fenrir-z replied:
added fixme

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
add fixme in cmd evaluate's --cis arg

phoenix-xhuang commented on ymir/command/mir/commands/evaluate.py:
```diff
@@ -18,12 +19,13 @@ def run(self) -> int:
                                          iou_thrs=self.args.iou_thrs,
-                                         need_pr_curve=self.args.need_pr_curve)
+                                         need_pr_curve=self.args.need_pr_curve,
+                                         cis=self.args.cis)
```
这里并不是cis，这里是class_names

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/commands/evaluate.py:
```diff
@@ -92,4 +96,11 @@ def bind_to_subparsers(subparsers: argparse._SubParsersAction, parent_parser: ar
                                      action='store_true',
                                      help='also generates pr curve in evaluation result')
+    # FIXME: using cis in prediction target
+    evaluate_arg_parser.add_argument('--cis',
```
--class_names

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -295,17 +165,14 @@ def evaluateImg(self, imgIdx: int, catId: int, aRng: Any, maxDet: int) -> Option
-
         if not len(ious) == 0:
             for tind, t in enumerate(p.iouThrs):
+                # for gind, g in enumerate(gt):
```
这里为啥要注释掉

fenrir-z replied:
现在使用 match_result 这个成员来记录匹配信息

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -328,15 +195,20 @@ def evaluateImg(self, imgIdx: int, catId: int, aRng: Any, maxDet: int) -> Option
-                gt_pb_index_id = d['cm'][tind, maxDet][1]
-                d['cm'][tind, maxDet] = (mirpb.ConfusionMatrixType.IGNORED, gt_pb_index_id)
+        # for dind, d in enumerate(dt):
+        #     if dtIg[tind, dind]:
```
注释掉的是？

fenrir-z replied:
同上

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+        self._asset_id_to_ordered_idxes = {asset_id: idx for idx, asset_id in enumerate(self._ordered_asset_ids)}
+        # ordered list of class / category ids
+        self._ordered_class_ids = sorted(list(keyword_to_idx.cis.keys()))
+        self._ck_idx: Dict[str, mirpb.AssetAnnoIndex] = {key: value for key, value in mir_keywords.ck_idx.items()}
```
为什么还有ck

fenrir-z replied:
removed

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+                 conf_thr: float,
+                 dataset_id: str,
+                 as_gt: bool,
+                 asset_ids: Iterable[str] = None) -> None:
```
asset_ids并没有地方传进来，去掉简化。有实际需求了再加

fenrir-z replied:
removed

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+            asset_ids (Iterable[str]): asset ids you want to include in MirDataset instance, None means include all
+        """
+        task_annotations = mir_annotations.ground_truth if as_gt else mir_annotations.prediction
+        keyword_to_idx = mir_keywords.gt_idx if as_gt else mir_keywords.pred_idx
```
mir_keywords传进来的作用就是为了得到_ordered_class_ids，完全可以去掉。外面传的时候不是MirAnnotations ，而是具体的SingleTaskAnnotations，同时as_gt这个参数也就不需要了。

现在参数列表冗余

phoenix-xhuang replied:
后面_get_annotations本身就要遍历annotation，过程中也可以得到class_ids

fenrir-z replied:
移除了 MirDataset 中冗余的部分

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+        self.img_cat_to_annotations: Dict[Tuple[int, int], List[dict]] = defaultdict(list)
+        annos = self._get_annotations(single_task_annotations=task_annotations,
+                                      asset_idxes=self.get_asset_idxes(),
+                                      class_ids=self.get_class_ids(),
```
这里的class_ids是多余，本身就是全量类别，等同于没有过滤

fenrir-z replied:
移除了此处的 class_ids 参数

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+
+        self.img_cat_to_annotations: Dict[Tuple[int, int], List[dict]] = defaultdict(list)
+        annos = self._get_annotations(single_task_annotations=task_annotations,
+                                      asset_idxes=self.get_asset_idxes(),
```
asset_idxes这个参数是多余的，逻辑上一直是全量，反而增加了匹配的计算理

fenrir-z replied:
移除了此处的 assset_idxes 参数

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+        annos = self._get_annotations(single_task_annotations=task_annotations,
+                                      asset_idxes=self.get_asset_idxes(),
+                                      class_ids=self.get_class_ids(),
+                                      conf_thr=None if as_gt else conf_thr)
```
conf_thr在外部就可以判断好，不需要在这里面来。MirDataset不需要知道自己是gt还是pred，不然就耦合起来了

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+                                  error_message='no annotations in evaluated dataset')
+
+        # ordered list of asset / image ids
+        self._ordered_asset_ids = sorted(asset_ids or mir_metadatas.attributes.keys())
```
mir_metadatas也不是必要的，需要的只是asset_ids。真正需要的只有一个SingleTaskAnnotations，甚至asset_ids都可以从SingleTaskAnnotations得到。既然是重构，就多思考下这里的业务逻辑和流程，有没有可以综合改进的地方，而不是纯重复完成以前的设计

fenrir-z replied:
移除了 MirDataset.__init__ 中 mir_metadatas, mir_keywords 参数

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_utils.py:
```diff
@@ -0,0 +1,256 @@
+from mir.protos import mir_command_pb2 as mirpb
+
+
+class MirDataset:
```
MirDataset这个类别的存在必要是什么，和SingleTaskAnnotations的差别是什么

fenrir-z replied:
原来在 coco_eval 里面有个类叫 COCO，负责读取标注文件和预测信息，为 evaluate 流程提供数据
在集成进来的时候，写了一个 MirCoco 类，来从 mir 的 pb 文件里面读取信息，相当于是 COCO 的替换

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_voc.py:
```diff
@@ -0,0 +1,275 @@
+    return see
+
+
+def det_evaluate(mir_dts: List[MirDataset], mir_gt: MirDataset, config: mirpb.EvaluateConfig) -> mirpb.Evaluation:
```
看上去整个业务流程可以非常简化，也不用创建一个MirDataset类了，就是一个det_evaluate函数。这里直接传SingleTaskAnnotations和config：
mir_dts: List[SingleTaskAnnotations], mir_gt: SingleTaskAnnotations, config: mirpb.EvaluateConfig

fenrir-z replied:
updated

phoenix-xhuang commented:
针对实际情况，有一本书推荐 https://book.douban.com/subject/3265691/
里面有介绍如何定义子系统去耦合：
S:Subject Area，表示子问题域，其核心思想是要通过业务来分解系统，尽量保证业务独立和低耦合。

这个evaluation重构是一个很好的例子，存在的问题是流程驱动的设计：把一个业务流程切成几段，并一段段实现。把所有过程中用到的metadatas, keywords, annotations放到一个mirdataset的类，在实际的evaluate中再使用这个mirdataset的类。流程驱动的设计问题是高耦合，做一个evaluate至始至终MirDataset都把“我是不是gt“这个信息揣在身上，这就是系统耦合。
这里缺少的是从业务独立和低耦合的角度思考：为什么需要这个mirdataset类？核心业务是evaluate，核心数据是SingleTaskAnnotations。其他的mir_repo/metadatas/keywords与SingleTaskAnnotations揉杂在一起只会影响拔剑的速度。先把这部分核心功能设计出来。而且是以高独立和低耦合的模式设计出来，后续再围绕这个核心完成整个的业务流程。

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
cmd evaluate: --cis changed to --class-names

Commit:
update MirDataset's __init__ func

Commit:
MirDataset: remove unuseed get_class_ids

Commit:
MirDataset: _get_annotations -> _aggregate_annotations

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
CocoDetEval: remove unused part

Commit:
use task_annotations (gt or pred) in function det_evaluate in det_eval_coco and det_eval_voc

Commit:
MirDataset -> MirCoco, remove unused field: _task_annotations

Commit:
remove unused code

Commit:
update

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -1,117 +1,70 @@
-                    continue
                 if conf_thr is not None and annotation.score < conf_thr:
                     continue
 
```
asset_idx并没有用到了

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -1,117 +1,70 @@
         annotation_idx = 1
-        for asset_idx in asset_idxes:
-            asset_id = self._ordered_asset_ids[asset_idx]
+        for asset_idx, asset_id in enumerate(self._ordered_asset_ids):
```
外面传进来的asset_ids并没有用，完全可以去掉：
外面设置asset_ids就是metadata的key，也就是数据集所有assets。换句话说，就是这里的single_task_annotations的key超集，而下面一行又判断如果不在image_annotations就跳过，这不就等同于从image_annotations取key吗？绕了这么一大圈，传了一个用不到的参数。这个参数还限制了gt与dt必须来自于同一个metadata

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -1,117 +1,70 @@
-            as_gt (bool): if false, use preds in mir_annotations and mir_keywords, if true, use gt
-            asset_ids (Iterable[str]): asset ids you want to include in MirCoco instance, None means include all
+            asset_ids (Collection[str]): asset ids (hashes)
+            pred_or_gt_annotations (mirpb.SingleTaskAnnotations): pred or gt annotations
```
s/pred_or_gt_annotations/task_annotations

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,80 @@
+from mir.protos import mir_command_pb2 as mirpb
+
+
+def det_evaluate(
```
以前商量过这个接口是要设计为通用evaluation：
`def det_evaluate(mir_root: str, gt_rev_tid: revs_parser.TypRevTid, pred_rev_tid: revs_parser.TypRevTid, conf_thr: float, iou_thrs: str, class_ids: List[int] = [], need_pr_curve: bool = False)`

fenrir-z replied:
updated

phoenix-xhuang replied:
这里并没有改变呢？可以趁这个机会把这个接口设计为跨dataset的evaluate使用，现在的单dataset情况下是一种gt_rev_tid和pred_rev_tid相同的特例

fenrir-z replied:
updated
没看到 gt_rev_tid 和 pred_rev_tid 是分开的

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,80 @@
+    evaluate_config.need_pr_curve = need_pr_curve
+    evaluate_config.gt_dataset_id = dataset_id
+    evaluate_config.pred_dataset_ids.append(dataset_id)
+    evaluate_config.class_ids[:] = class_ids or mir_keywords.gt_idx.cis.keys()
```
mir_metadatas 和 mir_keywords都可以省掉，只需要mir_annotations

fenrir-z replied:
updated

Commit:
det_eval_voc remove asset_ids

Commit:
det_evaluate: remove asset_ids

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
fix CI

Commit:
update

Commit:
update

phoenix-xhuang commented on ymir/command/.flake8:
```diff
@@ -1,5 +1,5 @@
 [flake8]
-ignore = C901, W503
+ignore = C901, W503, W504
```
W504会影响的地方是？

fenrir-z replied:
det_eval_voc.py L117

phoenix-xhuang replied:
看上去L117格式的确可以修改下，把乘号放到下一行？

fenrir-z replied:
试过了，照样会提示。
加 W504 会有啥影响吗？

fenrir-z replied:
把公式拆了

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -1,125 +1,69 @@
-            conf_thr (float): lower bound of annotation confidence score
-            as_gt (bool): if false, use preds in mir_annotations and mir_keywords, if true, use gt
-            asset_ids (Iterable[str]): asset ids you want to include in MirCoco instance, None means include all
+            asset_ids (Collection[str]): asset ids (hashes)
```
这一行可以去掉了

fenrir-z replied:
removed

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,74 @@
+) -> Tuple[mirpb.Evaluation, mirpb.MirAnnotations]:
+    mir_annotations: mirpb.MirAnnotations
+    mir_keywords: mirpb.MirKeywords
+    mir_annotations, mir_keywords = mir_storage_ops.MirStorageOps.load_multiple_storages(
```
这里如果判断pred_rev_tid与gt_rev_tid相同，则优化为只读取一次，这样接口扩展了而性能并不会受影响

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -24,11 +25,13 @@ def __init__(self,
                  evaluate_conf_thr: float = mir_settings.DEFAULT_EVALUATE_CONF_THR,
```
记得以前问过，现在还是不太明白这个类和proto里面的EvaluateConfig有什么区别，里面的字段基本是重合的，是否可以直接用EvaluateConfig呢？或者把EvaluateConfig作为这里的一个成员变量

fenrir-z replied:
removed build_config, now use evaluate_config

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,74 @@
+from mir.protos import mir_command_pb2 as mirpb
+
+
+def det_evaluate(
```
要反映上述设计，名字修改为det_evaluate_datasets

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -50,20 +53,19 @@ def __build_and_save(cls, mir_root: str, mir_datas: Dict['mirpb.MirStorage.V', A
                 conf_thr=build_config.evaluate_conf_thr,
                 iou_thrs=build_config.evaluate_iou_thrs,
                 need_pr_curve=build_config.evaluate_need_pr_curve,
+                class_ids=build_config.evaluate_class_ids or list(mir_keywords.gt_idx.cis.keys()),
```
det_evaluate_with_pb的接口设计为：”class_ids不存在则使用gt“的逻辑由外部调用者判断，这就造成了每个调用者都要记住这个规则。还是等于是把耦合逻辑散步在调用方。既然det_evaluate_with_pb知道哪个是gt，完全可以把这个逻辑放在det_evaluate_with_pb内部，外部要么传class_ids进来，要么就不传

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,74 @@
+    need_pr_curve: bool = False,
+    mode: str = 'voc',  # voc or coco
+) -> Tuple[mirpb.Evaluation, mirpb.MirAnnotations]:
+    # evaluation = mirpb.Evaluation()
```
这一行可以去掉

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -50,20 +53,19 @@ def __build_and_save(cls, mir_root: str, mir_datas: Dict['mirpb.MirStorage.V', A
                 and mir_annotations.prediction.image_annotations):
-            evaluation, _ = det_eval.det_evaluate_with_pb(
-                mir_metadatas=mir_metadatas,
+            evaluation, _ = det_eval_ops.det_evaluate_with_pb(
```
这里还隐含了一个业务逻辑：并没有使用返回的修改过的带cm的annotation，而是默认会把cm加到传入的annotation上。要么把这个逻辑摆到台面上讲明，然后只返回evaluation；要么使用返回的annotation结果替换掉原始annotation。看上去现在默认是前者，那就只保留第一个返回值

fenrir-z replied:
updated: returns only Evaluation

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
update

Commit:
update

Commit:
det_eval_ops.det_eval -> det_eval_datasets

Commit:
update

Commit:
remove MirStorageOpsBuildConfig

phoenix-xhuang commented on ymir/backend/src/ymir_viz/src/controllers/evaluation_controller.py:
```diff
@@ -45,11 +45,12 @@ def dataset_fast_evaluation(user_id: str, repo_id: str, branch_id: str, conf_thr
-                                          conf_thr=conf_thr,
-                                          iou_thrs=str(iou_thr),
-                                          need_pr_curve=need_pr_curve)
+    evaluation, _ = det_eval_ops.det_evaluate_datasets(mir_root=mir_root,
```
返回值是不是要把后面一部分去掉？

fenrir-z replied:
removed

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_ops.py:
```diff
@@ -0,0 +1,80 @@
+                                                                            ms=mirpb.MirStorage.MIR_ANNOTATIONS)
+    prediction = mir_annotations.prediction
+
+    evaluate_config = mirpb.EvaluateConfig()
```
这里可以直接调用create_evaluate_config?

fenrir-z replied:
不行，循环依赖了

phoenix-xhuang replied:
1.看上去mir_storage_ops和det_eval_ops依赖有点问题呢，det_eval_ops里面用到了mir_storage_ops.MirStorageOps，而mir_storage_ops里面用到了det_eval_ops. det_evaluate_with_pb。这样互相依赖不会报错吗？是否解决下更好？

2. create_evaluate_config就是一个helper函数，从参数填成一个proto message。除了对proto有依赖，不需要其他内容，这里应该解决循环依赖的问题，而不是手动重写一个完全重复的代码

fenrir-z replied:
主要是两个方法，一个是 det_eval_datasets，一个是 det_eval_with_pb
产生循环依赖的就是前面这个函数，我把它提到另外一个新建的文件里面去了

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_voc.py:
```diff
@@ -0,0 +1,273 @@
+                 config: mirpb.EvaluateConfig) -> mirpb.Evaluation:
+    if config.conf_thr < 0 or config.conf_thr > 1:
+        raise MirRuntimeError(error_code=MirCode.RC_CMD_INVALID_ARGS, error_message='invalid conf_thr')
+
```
后面的业务逻辑有用到，但这里缺了的一个检查：len(predictions) == len(config.pred_dataset_ids)
而且这里有一个索引的对应的关系，是否要考虑做得更简化些：比如config的pred_dataset_ids直接做成{dataset_id: prediction}的方式？避免有逻辑依赖性的数据分散在两个参数里

phoenix-xhuang replied:
方案一：把det_evaluate和det_evaluate_with_pb，设计为如下接口
predictions: Dict[str, mirpb.SingleTaskAnnotations], ground_truth: Tuple[str, mirpb.SingleTaskAnnotations,
                 config: mirpb.EvaluateConfig) -> mirpb.Evaluation:

然后把gt_dataset_id和pred_dataset_id，从mirpb.EvaluateConfig去掉

方案二：把ground_truth和predictions的引用，直接加到config: mirpb.EvaluateConfig

这两个方案目的都一样：把predictions这样与config.pred_dataset_ids有隐含顺序对应关系的规则，明面化

fenrir-z replied:
方案一吧，我改一下。

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -328,8 +327,14 @@ def save_and_commit(cls,
+        copied_evaluate_config = mirpb.EvaluateConfig()
+        copied_evaluate_config.CopyFrom(evaluate_config)
+
+        if not copied_evaluate_config.gt_dataset_id and not copied_evaluate_config.pred_dataset_ids:
```
这里看上去应该强制重写copied_evaluate_config，把gt_dataset_id和pred_dataset_ids都强制设定为revs_parser.join_rev_tid(mir_branch, task.task_id)

否则的话，外面传入的dataset_id会与底层数据不一致

phoenix-xhuang replied:
甚至这里应该做个检查，应该保证传进来的两个dataset_id字段为空

fenrir-z replied:
这里分两种情况讨论：

1. 如果是 evaluate 命令，或者由 fast_evaluate 接口传过来，copied_evaluate_config 里面的 gt 和 pred dataset id 应该是 src_rev_tid，这一点由命令的发起者保证
2. 如果是其他命令，例如 filter，copied_evaluate_config 里面的 gt 和 pred dataset id 应该是 dst_rev_tid，表示 evaluate 是基于 dst_rev_tid 所对应的数据集做出的

phoenix-xhuang replied:
fast_evaluate这样的命令，不会走mir_stoage_ops这条路，只需要evaluation的结果，不需要保存mir pb

不要按调用命令做出不同行为，应该是设计好底层行为，由调用命令自行使用。底层功能不要去考虑上层调用者的逻辑，否则又是拎不清逻辑。回到这里的save_and_commit，这个函数目的就是生成一个dst_rev_tid=mir_branch@task.task_id的数据集，你说gt和pred是不是应该等于dst_rev_tid

config里的gt/pred dataset_id是用于定义返回的Evaluation的key的，不是定义src的，设计上不要混一起了

phoenix-xhuang replied:
核心逻辑：evaluation_config里面的gt/pred dataset_id压根不会影响到要保存的mir_datas，只会影响到返回的Evaluation里面的key

fenrir-z replied:
我现在的逻辑是这样的：

1. 如果是 evaluate 命令，config 里面的 gt 和 pred dataset id 已经由 evaluate.py L56 定义好（是 src_rev_tid），返回的结果里面，evaluation.dataset_evaluations 里面也会有一个以 src_rev_tid 为 key 的结果
2. 如果是 fast_evaluate 接口，config 里面的 gt 和 pred dataset id 已经由 evaluation_controller.py L49 定义好（是 src_rev_tid），返回的结果里面，也会有一个以 src_rev_tid 为 key 的结果
3. 如果是其他命令，例如 copy 或 filter，由于外部没指定 config 的 gt / pred dataset id，所以会进入 mir_storage_ops.py L337，因为此时 evaluate 结果是根据结果数据集生成的，返回的结果里面，也会有一个以 dst_rev_tid 为 key 的结果

如果按照你的做法，强制让它们等于 dst_rev_tid，这和我的逻辑的区别在于：对于一个 mir evaluate 命令，其结果的 evaluation.dataset_evaluations 的 key，应该是 src 还是 dst

在这一点上，我认为 cmd evaluate 的行为应该大体上保持和 fast_evaluate 的行为一致。既然 fast_evaluate 结果的 key 是 src，那 cmd evaluate 的结果的 key 也应该是 src

phoenix-xhuang replied:
config的意义是在于：确定返回的Evaluation结构体里面的key的值，并不需要知道这个是src还是dst。要把key和src/dst这两个逻辑拆开。这个PR的review反复在介绍两点：抽象独立业务逻辑，确定模块边界。

从这里例子来说：det_evaluate_with_pb做为evaluate任务具体执行函数，只需要按config里面的dataset_id作为返回值的key。这就是边界，确定了以后往上推。现在det_evaluate_with_pb函数的调用者有两个，一是fast_evaluate的det_evaluate_datasets，这里因为不需要保存mir data，所以config里面用的原src id；另一个是storage_ops里的save_and_commit，只要走这里，目标都是保存mir data为一个版本。如果storage_ops这里不使用dst作为key，就会出现一个问题，那就是使用这个evaluate结果的时候，传入的key并不是这个版本的id。

模块边界要独立，就不是直接从调用者的角度来看问题，而是先把核心被调用函数的功能设计清楚。然后再从调用者的角度来想怎么调用，这里的两个调用者（storage_ops和det_evaluate_datasets）就分别应该准备好数据，然后把需要结果设置config里面的key指定（前者为dst，后者为src）。从这个逻辑就可以很容易推出：mir evaluate命令调用走的storage_ops路线，和fast_evaluate走的调用路线不一样，所以mir evaluate里面的key应该是dst。

phoenix-xhuang replied:
差异还是在于mir evaluate这个命令，因为现在没有具体的需求，所以只能按各自的理解来设计这个命令的行为。在我看来，mir evaluate要么不保存结果只做类似于mir log这样的显示，这样的话直接用src id是没问题的；但如果要保存结果，就意味着使用结果的场景会与产生结果时不一样，那这时候应该是dst id，否则使用这个数据的人还要知道来源的dataset是哪个。

从反推法：如果认为上面的后一种逻辑成立（保存的结果应该以src id为key），就可以推出来现在mirstorage里的evaluate key应该全部保存为src的id，但这是不合理的，毕竟像cm这些数据都是存在dst的annotation里面。

fenrir-z replied:
先改了。
这一点我仍然持保留意见。

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -13,29 +13,33 @@
-        self.evaluate_iou_thrs: str = evaluate_iou_thrs
-        self.evaluate_need_pr_curve: bool = evaluate_need_pr_curve
-        self.evaluate_src_dataset_id: str = evaluate_src_dataset_id
+def create_evaluate_config(conf_thr: float = mir_settings.DEFAULT_EVALUATE_CONF_THR,
```
这里的pred_dataset_ids接口，为什么设置为gt和pred要来自于同一个src_dataset_id？

fenrir-z replied:
updated

Commit:
update

Commit:
update

Commit:
updated

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
update

phoenix-xhuang commented on ymir/backend/src/ymir_controller/controller/invoker/invoker_cmd_evaluate.py:
```diff
@@ -36,11 +36,12 @@ def invoke(self) -> backend_pb2.GeneralResp:
-                                              conf_thr=ec.conf_thr,
-                                              iou_thrs=ec.iou_thrs_interval,
-                                              need_pr_curve=ec.need_pr_curve)
+        evaluation, _ = det_eval_ops.det_evaluate_datasets(mir_root=self._repo_root,
```
这里还有个多余返回值没有改

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/commands/evaluate.py:
```diff
@@ -92,4 +97,11 @@ def bind_to_subparsers(subparsers: argparse._SubParsersAction, parent_parser: ar
                                      dest='need_pr_curve',
                                      action='store_true',
                                      help='also generates pr curve in evaluation result')
+    # FIXME: using cis in prediction target
```
mir evaluate命令会在哪里用到？

phoenix-xhuang replied:
是否还有存在的必要？

fenrir-z replied:
对于命令行用户来说，这条命令还是有用的。
这个 FIXME 的意思是说，如果明确传入了 --class-names，就用 --class-names 的设置，否则用 prediction target（虽然现在还没有）

phoenix-xhuang replied:
明白你的意思，行吧，保留命令行功能的完整性

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
break dependency circle

fenrir-z committed:
update

fenrir-z committed:
update

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -13,29 +13,34 @@
+    evaluate_config.iou_thrs_interval = iou_thrs
+    evaluate_config.need_pr_curve = need_pr_curve
+    evaluate_config.class_ids[:] = class_ids
+    evaluate_config.gt_dataset_id = gt_dataset_id
```
前面商量是把config里面的gt/pred的dataset_id可以去掉，直接放到evaluate_pb函数的参数里？看到参数里已经在使用了，那这里的config里面字段是不是可以不用了

fenrir-z replied:
这样会让 evaluate_pb 多一个参数。
而且以前这样干过，后来 code review 的时候改回来了。

phoenix-xhuang replied:
那现在就确定下，是用config还是用pb参数？不应该两个地方都需要设置吧？

phoenix-xhuang replied:
照前面的建议，如果不希望增加参数，可以直接把现在gt的输入，从SingleTaskAnnotation，变成Tuple[str, SingleTaskAnnotation]就可以了

fenrir-z replied:
config 里面的 gt 和 pred dataset id 对结果是有说明作用的，保留下来比较好。

phoenix-xhuang replied:
所以调用evaluate_pb的时候，要填config里面的pred/gt dataset id，然后也要把这两个相同的内容传到参数里？这合理吗？

要觉得config里面的字段有用，就在evaluate_pb内部按指定格式补上好了

fenrir-z replied:
evaluate_config 里面的 pred / gt dataset id 根据传入参数指定

fenrir-z replied:
正如上午讨论的：移除了 evaluate_config 中的 pred / gt dataset id

phoenix-xhuang commented on ymir/command/mir/tools/mir_storage_ops.py:
```diff
@@ -50,20 +55,16 @@ def __build_and_save(cls, mir_root: str, mir_datas: Dict['mirpb.MirStorage.V', A
                                  mir_keywords=mir_keywords)
         mir_datas[mirpb.MirStorage.MIR_KEYWORDS] = mir_keywords
 
-        det_eval.reset_default_confusion_matrix(task_annotations=mir_annotations.prediction,
```
把你这里的隐含逻辑摆到台面上来：

如果要使用det_evaluate_with_pb函数，要注意：
1. EvaluateConfig可能会被改变
2. 调用时要确保ground_truth和prediction不为空
3. 自己要把传入进来的cm初始化为NotSet

从设计上来说，det_evaluate_with_pb的调用者（mir_storage_ops）就只是要做个evaluation，为什么需要知道这么多具体的evaluation逻辑，这不就是没有封装好？每一个调用者都要清楚这么多细节，才能正常使用，这样的设计就应该改进。更好的设计是：
1. det_evaluate_with_pb要改变config前内部copy一份
2. 对ground_truth和prediction是否为空的判断放到封装的函数内部，确定清楚要是任意为空会如何，不会空会如何
3. cm初始化为NotSet也封装在evaluation内部

调用者就传pred/gt/config（只包含conf这些）就行了，要的就是一个evaluation结果，作好封装，别让调用者参与到具体的业务逻辑里面来

fenrir-z replied:
> EvaluateConfig可能会被改变

严格说来，这个不是 evaluate_with_pb 的问题，而是默认参数的问题。
就算我把 copy 操作移到 evaluate_with_pb 里面，将来某个人在 mir_storage_ops::commit_and_save 函数里面改变了默认参数 evaluate_config 的值，还是会有问题
这里我只能把原因写在注释里面，单纯地把 copy 操作移到 evaluate_with_pb 里面是解决不了问题的

fenrir-z replied:
其他两项：cm 初始化为 NotSet，以及 ground_truth 和 prediction 为空的特殊情况处理，已经移入 evaluate 相关的函数里面。

fenrir-z replied:
对 evaluate_config 复制的行为做出了更改，不再采用复制操作

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
update

fenrir-z committed:
create_evaluate_config: remove unused args

Commit:
remove cmd: evaluate;
det_eval_ops: multiple predictions changed to single prediction;

Commit:
update

Commit:
viz: get_dataset_evaluations -> get_dataset_evaluation;
remove unused fields from mir_command.proto;
ctl: remove evaluate invoker;
cmd: remove evaluate cmd;

fenrir-z committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl

Commit:
update test case for mir_det_eval_ops

Commit:
add invoker_cmd_evaluate and it's test case back

Commit:
update

Commit:
update

Commit:
update

Commit:
backend proto: add CMD_EVALUATE back

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -697,133 +515,44 @@ def __init__(self) -> None:
+def det_evaluate(prediction: mirpb.SingleTaskAnnotations, ground_truth: mirpb.SingleTaskAnnotations,
+                 config: mirpb.EvaluateConfig) -> mirpb.Evaluation:
+    if len(ground_truth.image_annotations) == 0 or len(prediction.image_annotations) == 0:
+        return mirpb.Evaluation()
```
这里应该把下面的
    evaluation = mirpb.Evaluation()
    evaluation.config.CopyFrom(config)
放最上面，不然返回的结果里面没有config信息

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/backend/src/ymir_controller/controller/utils/invoker_mapping.py:
```diff
@@ -1,7 +1,7 @@
 from controller.invoker import (
     invoker_cmd_branch_commit,
-    invoker_cmd_evaluate,
```
放错位置了

fenrir-z replied:
updated

phoenix-xhuang commented on ymir/command/mir/tools/det_eval_coco.py:
```diff
@@ -697,133 +515,44 @@ def __init__(self) -> None:
-        for annotation in image_annotations.annotations:
-            annotation.cm = cm
-            annotation.det_link_id = -1
+    det_eval_utils.write_confusion_matrix(gt_annotations=ground_truth,
```
coco是没有iouThrs循环，也没有per class_ids的循环，这里是不需要吗？

fenrir-z replied:
coco 的话，所有 iou thrs 和 class ids 都一次性算好了，不需要在这里做循环。

Commit:
update

Commit:
update

phoenix-xhuang committed:
Merge branch 'dev' into zzw/refactor-evaluate-impl
